--- a/net/core/skbuff.c	2022-05-18 11:03:02.361259000 +0800
+++ b/net/core/skbuff.c	2022-05-18 11:02:09.832117270 +0800
@@ -82,6 +82,8 @@
 #include "datagram.h"
 #include "sock_destructor.h"
 
+#include <net/ra_nat.h>
+
 struct kmem_cache *skbuff_head_cache __ro_after_init;
 static struct kmem_cache *skbuff_fclone_cache __ro_after_init;
 #ifdef CONFIG_SKB_EXTENSIONS

@@ -1665,7 +1673,15 @@
 	memcpy((struct skb_shared_info *)(data + size),
 	       skb_shinfo(skb),
 	       offsetof(struct skb_shared_info, frags[skb_shinfo(skb)->nr_frags]));
-
+	memcpy(data, skb->head, FOE_INFO_LEN);
	
 	/*
 	 * if shinfo is shared we must drop the old head gracefully, but if it
 	 * is not we can just drop the old head and let the existing refcount 

--- a/include/linux/mdio.h	2022-03-28 15:57:11.000000000 +0800
+++ b/include/linux/mdio.h	2022-12-27 16:18:35.806004570 +0800
@@ -5,7 +5,7 @@
  */
 #ifndef __LINUX_MDIO_H__
 #define __LINUX_MDIO_H__
-
+#include <linux/bitfield.h>
 #include <uapi/linux/mdio.h>
 #include <linux/mod_devicetable.h>
 
@@ -14,8 +14,10 @@
  */
 #define MII_ADDR_C45		(1<<30)
 #define MII_DEVADDR_C45_SHIFT	16
+#define MII_DEVADDR_C45_MASK	GENMASK(20, 16)
 #define MII_REGADDR_C45_MASK	GENMASK(15, 0)
 
+
 struct gpio_desc;
 struct mii_bus;
 struct reset_control;
@@ -342,6 +344,16 @@
 	return MII_ADDR_C45 | devad << MII_DEVADDR_C45_SHIFT | regnum;
 }
 
+static inline u16 mdiobus_c45_regad(u32 regnum)
+{
+	return FIELD_GET(MII_REGADDR_C45_MASK, regnum);
+}
+
+static inline u16 mdiobus_c45_devad(u32 regnum)
+{
+	return FIELD_GET(MII_DEVADDR_C45_MASK, regnum);
+}
+
 static inline int __mdiobus_c45_read(struct mii_bus *bus, int prtad, int devad,
 				     u16 regnum)
 {


