--- a/include/linux/mdio.h
+++ b/include/linux/mdio.h
@@ -5,7 +5,7 @@
  */
 #ifndef __LINUX_MDIO_H__
 #define __LINUX_MDIO_H__
-
+#include <linux/bitfield.h>
 #include <uapi/linux/mdio.h>
 #include <linux/mod_devicetable.h>
 
@@ -14,8 +14,10 @@
  */
 #define MII_ADDR_C45		(1<<30)
 #define MII_DEVADDR_C45_SHIFT	16
+#define MII_DEVADDR_C45_MASK	GENMASK(20, 16)
 #define MII_REGADDR_C45_MASK	GENMASK(15, 0)
 
+
 struct gpio_desc;
 struct mii_bus;
 struct reset_control;
@@ -342,6 +344,16 @@
 	return MII_ADDR_C45 | devad << MII_DEVADDR_C45_SHIFT | regnum;
 }
 
+static inline u16 mdiobus_c45_regad(u32 regnum)
+{
+	return FIELD_GET(MII_REGADDR_C45_MASK, regnum);
+}
+
+static inline u16 mdiobus_c45_devad(u32 regnum)
+{
+	return FIELD_GET(MII_DEVADDR_C45_MASK, regnum);
+}
+
 static inline int __mdiobus_c45_read(struct mii_bus *bus, int prtad, int devad,
 				     u16 regnum)
 {


